### DISCUSSION TITLE
How do you move a Winter CMS site from a local server to a production server?

### DISCUSSION THREAD

**Discussion by spanjaan** [wintercms/winter#268] [Q&A] [answered]
Hello,

i am new in winter cms. i haven't any idea for migration.
please provide me full course(step by step) how to migrate local server to web server.

Thank You.

**Comment by RomainMazB** [comment]
There is not that much to do, but it depends on your webserver.

### If you can run composer on it
- Using an ftp client, send all the files from the local installation to the webserver, except the `modules` directory and the `vendors` one (you can also skip the plugins that are listed inside the `composer.json` file but this won't make you win that much time if none of them includes a "nested" `vendors` directory)
- run `composer install` to download the modules and vendors (this is way faster than sending them through ftp and this make sure they don't contain any debugging statement you may have introduced inside those files during development)
- configure the `.env` file or config files for the database connection
- run `php artisan winter:up` to run the migrations. **This command will create an admin with a random password, make sure to copy it before moving on to the next steps**
- at this point the website should run

If there are many database data that you need to transfer, just export/import the data  (without the tables structures) through phpmyadmin or equivalent.

_____
### If you can't run `composer` or `php` commands on the webserver

The basic steps for any cms would work: send all the file through ftp, wait three hours it ends and then migrate all the database including structure (you can migrate the database during the file transfer, but you will still need to wait 2h50 then).

In that case the admin login and password remains the same that you used on your local machine during development, so make sure to not keep admin/admin ;)

**Comment by LukeTowers** [comment, marked as answer]
## Quick and dirty
Use this method if you're not comfortable using command line interfaces or working with CI / CD techniques / platforms:

1. Zip all of the project files into a single zipped file (make sure you include the hidden files and folders like `.htaccess`, `.env`, `.gitignore`, `.git`, etc.
2. Upload the zipped folder to the webroot on your server (i.e. `public_html`)
3. Unzip the folder's contents into the webroot on your server
4. If you're using a database (and it wasn't already migrated; i.e. SQLite stored as a file in the `storage` folder), migrate the database to the server using whatever methods you're most comfortable with and change the configuration values in either `.env` or `config/database.php` to point to the new database.
5. Make sure the following configuration values are set correctly:
- [ ] `config/app.php` -> `'debug' => false` (or `APP_DEBUG=false` if using a `.env` file)
- [ ] `config/app.php` -> `'url' => 'https://example.com'` (or `APP_URL="https://example.com"` if using a `.env` file) (use the publicly accessible URL for the site on your server instead of `example.com`
- [ ] verify `config/cms.php` -> `'restrictBaseDir' => true` This should never be `false` in a production environment
- [ ] (optional) `config/cms.php` -> `'databaseTemplates' => true` Will store future changes to the theme structure and content in the database so that re-uploading your files won't overwrite site content.

## Automated Deployments via CI/CD

This could be a whole article in and of itself but the basics boil down the following:

### Project structure
- Use source control to keep track of your project
- Use Composer to manage your dependencies
- Use a [public directory](https://wintercms.com/docs/setup/configuration#public-folder)
- Ignore the `modules` & `vendor` directories, along with your public folder and any plugins you include via Composer.
- Use the [`.env` file](https://wintercms.com/docs/setup/configuration#dotenv-configuration) to manage your secrets (never store secrets in source control!), ensuring that each environment has the correct configuration
- Use the [`cms.databaseTemplates` feature](https://wintercms.com/docs/cms/themes#database-driven-themes) to protect any changes to theme content files from being clobbered between deployments

### CI / CD Recommendations
- Have some way to trigger deployments (I typically just use GitHub webhooks)
- Use [atomic deployments](https://www.netlify.com/blog/2021/02/23/terminology-explained-atomic-and-immutable-deploys/) (Note: make sure the storage location is shared between deployments, otherwise you'll run into issues with new deployments not having access to files that were uploaded on an old deployment

### Deployment Steps
- Copy current state of repository to the deployment folder (including necessary environment variables and secrets in a `.env` file)
- Install dependencies with `composer install --no-dev`
- Trigger package discovery with `php artisan package:discover`
- Enable maintenance mode with `php artisan down`
- Run any pending migrations with `php artisan winter:up`
- Bust the cache to prevent caching issues between deployments with `php artisan winter:version && php artisan cache:clear`
- Regenerate the public folder with `php artisan winter:mirror public --relative`
- Disable maintenance mode with `php artisan up`

### Some CI / CD services for your reference
- https://envoyer.io/
- https://github.com/features/actions
- https://deploybot.com/

**Reply by spanjaan** [reply]
@LukeTowers thank you so much for your valuable reply.

Please share me some idea about push wintercms locale to web server using github actions. (some docs, tutorial etc.)
Thank you.

**Reply by LukeTowers** [reply]
There's lots of resources for doing it for Laravel projects (i.e. https://www.amezmo.com/laravel-hosting-guides/deploying-laravel-with-github-actions as the first google result for it) so you should be able to pick one that works for your target hosting environment and then just make minor tweaks for Winter instead of Laravel. Let me know if you have any more specific questions.
